# 🏦 Risco de Crédito Global — Pipeline de Dados

**Disciplina:** Ciência de Dados — UniCEUB
**Objetivo:** transformar dados brutos de crédito em uma base analítica pronta para Business Intelligence.

Este notebook produz três arquivos em `output/`, consumidos pelo painel do Power BI:

| Arquivo | Conteúdo | Grão |
|---|---|---|
| `credit_clean.csv` | carteira de clientes tratada + features | 1 linha = 1 cliente |
| `bcb_macro.csv` | séries macroeconômicas do Banco Central | 1 linha = 1 mês |
| `kpis_*.csv` | indicadores agregados para os visuais | variado |

> ⚠️ **Nota metodológica.** A carteira de clientes vem do dataset *Give Me Some Credit*
> (Kaggle, 2011 — consumidores **norte-americanos**, renda em USD). As séries do BCB são
> **macroeconomia brasileira**. São populações distintas: o BCB entra como **contexto setorial**,
> não como variável explicativa da carteira. As duas bases não se relacionam por chave —
> o painel as trata em seções separadas. Ver ADR-001 em `docs/ADR.md`.

**Ordem do pipeline** (a ordem importa — ver ADR-007):
`carregar → deduplicar → sentinelas → imputar renda → recalcular razão → winsorizar → features → exportar`

---
## 0. Setup

In [1]:
%pip install pandas numpy requests --quiet

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import os, zipfile
import numpy as np
import pandas as pd
import requests

pd.set_option('display.width', 120)
os.makedirs('output', exist_ok=True)
os.makedirs('data', exist_ok=True)

print('Ambiente pronto. pandas', pd.__version__)

Ambiente pronto. pandas 3.0.1


---
## 1. Dados macroeconômicos — API do Banco Central (SGS)

O BCB expõe séries temporais via API pública. Cada série tem um código numérico.

**Séries utilizadas** (nomes conferidos no Portal de Dados Abertos do BCB):

| Código | Nome oficial | Unidade |
|---|---|---|
| 21082 | Inadimplência da carteira de crédito — Total | % |
| 20631 | Concessões de crédito — Total | R$ milhões |
| 20714 | Taxa média de juros das operações de crédito — Total | % a.a. |

> A série **25433** (taxa média *mensal* de juros) foi **descartada**: é a mesma variável
> da 20714 em outra periodicidade (correlação 0,9999). Manter as duas duplicaria a informação
> no modelo. Ver ADR-002.

In [3]:
def baixar_serie_bcb(codigo, nome, inicio='01/01/2015', fim='01/12/2024'):
    """Baixa uma série do SGS/BCB e devolve um DataFrame [data, <nome>].

    codigo : int  — código da série no SGS
    nome   : str  — nome da coluna de valores no DataFrame resultante
    """
    url = (f'https://api.bcb.gov.br/dados/serie/bcdata.sgs.{codigo}/dados'
           f'?formato=json&dataInicial={inicio}&dataFinal={fim}')

    resp = requests.get(url, timeout=30)
    resp.raise_for_status()          # levanta exceção se o status não for 2xx

    df = pd.DataFrame(resp.json())
    df['data']  = pd.to_datetime(df['data'], format='%d/%m/%Y')
    df['valor'] = pd.to_numeric(df['valor'], errors='coerce')
    df = df.rename(columns={'valor': nome})

    print(f'  serie {codigo:>6} | {nome:<28} | {len(df):>4} registros')
    return df


SERIES = [
    (21082, 'Inadimplencia_Pct'),          # %
    (20631, 'Concessoes_Credito_RS_Mi'),   # R$ milhões
    (20714, 'Juros_Medio_Aa_Pct'),         # % ao ano
]

print('Baixando séries do BCB...')
frames = [baixar_serie_bcb(cod, nome) for cod, nome in SERIES]

Baixando séries do BCB...


  serie  21082 | Inadimplencia_Pct            |  120 registros


  serie  20631 | Concessoes_Credito_RS_Mi     |  120 registros


  serie  20714 | Juros_Medio_Aa_Pct           |  120 registros


In [4]:
# Consolidar as séries num único DataFrame mensal
bcb = frames[0]
for f in frames[1:]:
    bcb = bcb.merge(f, on='data', how='outer')

bcb = bcb.sort_values('data').reset_index(drop=True)

# Registrar quando os dados foram extraídos — as séries do BCB sofrem revisão
# retroativa, então a data de extração faz parte da procedência do dado (ADR-003)
bcb['Data_Extracao'] = pd.Timestamp.today().normalize()

print(f'\nDataset macro consolidado: {bcb.shape[0]} meses x {bcb.shape[1]} colunas')
print(f'Período: {bcb["data"].min():%Y-%m} a {bcb["data"].max():%Y-%m}')
print(f'Valores ausentes: {bcb.isnull().sum().sum()}')
bcb.tail()


Dataset macro consolidado: 120 meses x 5 colunas
Período: 2015-01 a 2024-12
Valores ausentes: 0


,data,Inadimplencia_Pct,Concessoes_Credito_RS_Mi,Juros_Medio_Aa_Pct,Data_Extracao
115,2024-08-01,3.33,627050,27.59,2026-08-22
116,2024-09-01,3.34,639146,27.49,2026-08-22
117,2024-10-01,3.30,649340,27.94,2026-08-22
118,2024-11-01,3.27,620293,28.45,2026-08-22
119,2024-12-01,3.08,699427,28.50,2026-08-22


In [5]:
# encoding='utf-8-sig' preserva acentos ao abrir no Power BI / Excel (ADR-009)
bcb.to_csv('output/bcb_macro.csv', index=False, encoding='utf-8-sig')
print(f'output/bcb_macro.csv  <-  {bcb.shape[0]} linhas x {bcb.shape[1]} colunas')

output/bcb_macro.csv  <-  120 linhas x 5 colunas


---
## 2. Carteira de clientes — dataset Give Me Some Credit

In [6]:
with zipfile.ZipFile('GiveMeSomeCredit.zip') as z:
    z.extract('cs-training.csv', path='data/')

df_raw = pd.read_csv('data/cs-training.csv', index_col=0)
print(f'Base bruta: {df_raw.shape[0]:,} linhas x {df_raw.shape[1]} colunas')

Base bruta: 150,000 linhas x 11 colunas


In [7]:
COLUNAS_PT = {
    'SeriousDlqin2yrs'                     : 'Inadimplente',
    'RevolvingUtilizationOfUnsecuredLines' : 'Utilizacao_Credito',
    'age'                                  : 'Idade',
    'NumberOfTime30-59DaysPastDueNotWorse' : 'Atrasos_30_59dias',
    'DebtRatio'                            : 'Razao_Divida',
    'MonthlyIncome'                        : 'Renda_Mensal',
    'NumberOfOpenCreditLinesAndLoans'      : 'Linhas_Credito_Abertas',
    'NumberOfTimes90DaysLate'              : 'Atrasos_90dias',
    'NumberRealEstateLoansOrLines'         : 'Emprestimos_Imoveis',
    'NumberOfTime60-89DaysPastDueNotWorse' : 'Atrasos_60_89dias',
    'NumberOfDependents'                   : 'Dependentes',
}

# Falha alto se o dataset de origem mudar de esquema, em vez de renomear silenciosamente
faltando = set(COLUNAS_PT) - set(df_raw.columns)
assert not faltando, f'Colunas ausentes na base bruta: {faltando}'

df_raw = df_raw.rename(columns=COLUNAS_PT)
print('Colunas traduzidas:', list(df_raw.columns))

Colunas traduzidas: ['Inadimplente', 'Utilizacao_Credito', 'Idade', 'Atrasos_30_59dias', 'Razao_Divida', 'Renda_Mensal', 'Linhas_Credito_Abertas', 'Atrasos_90dias', 'Emprestimos_Imoveis', 'Atrasos_60_89dias', 'Dependentes']


In [8]:
# Diagnóstico da base bruta — o que precisa ser tratado
diagnostico = pd.DataFrame({
    'Tipo'    : df_raw.dtypes,
    'Nulos'   : df_raw.isnull().sum(),
    'Nulos_%' : (df_raw.isnull().mean() * 100).round(2),
    'Minimo'  : df_raw.min(numeric_only=True),
    'Mediana' : df_raw.median(numeric_only=True).round(2),
    'Maximo'  : df_raw.max(numeric_only=True),
})
print(diagnostico.to_string())
print(f'\nLinhas duplicadas : {df_raw.duplicated().sum():,}')
print(f'Taxa de inadimplência : {df_raw["Inadimplente"].mean()*100:.2f}%')

                           Tipo  Nulos  Nulos_%  Minimo  Mediana     Maximo
Inadimplente              int64      0     0.00     0.0     0.00        1.0
Utilizacao_Credito      float64      0     0.00     0.0     0.15    50708.0
Idade                     int64      0     0.00     0.0    52.00      109.0
Atrasos_30_59dias         int64      0     0.00     0.0     0.00       98.0
Razao_Divida            float64      0     0.00     0.0     0.37   329664.0
Renda_Mensal            float64  29731    19.82     0.0  5400.00  3008750.0
Linhas_Credito_Abertas    int64      0     0.00     0.0     8.00       58.0
Atrasos_90dias            int64      0     0.00     0.0     0.00       98.0
Emprestimos_Imoveis       int64      0     0.00     0.0     1.00       54.0
Atrasos_60_89dias         int64      0     0.00     0.0     0.00       98.0
Dependentes             float64   3924     2.62     0.0     0.00       20.0



Linhas duplicadas : 609
Taxa de inadimplência : 6.68%


---
## 3. Limpeza

A ordem das etapas é deliberada (ADR-007): tudo que distorce estatísticas de resumo
(duplicatas, códigos sentinela, outliers) é tratado **antes** da imputação, para que as
medianas usadas na imputação não sejam contaminadas.

In [9]:
df = df_raw.copy()
auditoria = []   # registro de cada etapa, exportado ao final

def registrar(etapa, detalhe):
    auditoria.append({'Etapa': etapa, 'Detalhe': detalhe, 'Linhas_Restantes': len(df)})
    print(f'{etapa:<32} {detalhe:<44} -> {len(df):,} linhas')

In [10]:
# --- 3.1 Duplicatas -------------------------------------------------------
n = len(df)
df = df.drop_duplicates()
registrar('Duplicatas removidas', f'{n - len(df):,} linhas idênticas')

# --- 3.2 Idade inválida ---------------------------------------------------
n = len(df)
df = df[df['Idade'] > 0]
registrar('Idade inválida removida', f'{n - len(df):,} registro(s) com idade = 0')

Duplicatas removidas             609 linhas idênticas                         -> 149,391 linhas
Idade inválida removida          1 registro(s) com idade = 0                  -> 149,390 linhas


In [11]:
# --- 3.3 Códigos sentinela nas colunas de atraso --------------------------
# 96 e 98 não são contagens: são códigos especiais do dataset original.
# Tratá-los como "98 atrasos" inflaria qualquer agregação (ADR-004).
ATRASOS = ['Atrasos_30_59dias', 'Atrasos_60_89dias', 'Atrasos_90dias']

marcados = (df[ATRASOS] >= 90).any(axis=1).sum()
df['Registro_Sentinela'] = (df[ATRASOS] >= 90).any(axis=1).astype(int)

for col in ATRASOS:
    df.loc[df[col] >= 90, col] = np.nan
    df[col] = df[col].fillna(df[col].median())

registrar('Sentinelas 96/98 tratadas', f'{marcados} registros, mediana imputada')
print(f'  máximo por coluna após tratamento: '
      f'{ {c: int(df[c].max()) for c in ATRASOS} }')

Sentinelas 96/98 tratadas        225 registros, mediana imputada              -> 149,390 linhas
  máximo por coluna após tratamento: {'Atrasos_30_59dias': 13, 'Atrasos_60_89dias': 11, 'Atrasos_90dias': 17}


In [12]:
# --- 3.4 Imputação da renda ----------------------------------------------
# Mediana por faixa etária preserva melhor a estrutura da variável do que a
# mediana global — renda tem relação forte com idade.
df['Renda_Imputada'] = df['Renda_Mensal'].isnull().astype(int)
nulos_antes = df['Renda_Imputada'].sum()

_faixa = pd.cut(df['Idade'], bins=[0, 30, 45, 60, 200])
df['Renda_Mensal'] = (df.groupby(_faixa, observed=True)['Renda_Mensal']
                        .transform(lambda s: s.fillna(s.median())))
df['Renda_Mensal'] = df['Renda_Mensal'].fillna(df['Renda_Mensal'].median())

registrar('Renda imputada', f'{nulos_antes:,} nulos ({nulos_antes/len(df)*100:.1f}%) por faixa etária')

Renda imputada                   29,221 nulos (19.6%) por faixa etária        -> 149,390 linhas


In [13]:
# --- 3.5 Recálculo da Razão Dívida/Renda ---------------------------------
# No dataset original, quando MonthlyIncome é nulo o campo DebtRatio não guarda
# uma razão, e sim a DÍVIDA ABSOLUTA (mediana 1.159 contra 0,296 nos demais).
# Com a renda já imputada, a razão pode ser reconstruída (ADR-005).
mask = df['Renda_Imputada'] == 1
print(f'Antes  | mediana onde a renda era ausente: {df.loc[mask, "Razao_Divida"].median():>10,.2f}')

df.loc[mask, 'Razao_Divida'] = (df.loc[mask, 'Razao_Divida']
                                / df.loc[mask, 'Renda_Mensal'].replace(0, np.nan))
df['Razao_Divida'] = df['Razao_Divida'].fillna(df['Razao_Divida'].median())

print(f'Depois | mediana onde a renda era ausente: {df.loc[mask, "Razao_Divida"].median():>10,.2f}')
registrar('Razão dívida/renda recalculada', f'{mask.sum():,} registros reconstruídos')

Antes  | mediana onde a renda era ausente:   1,198.00
Depois | mediana onde a renda era ausente:       0.22
Razão dívida/renda recalculada   29,221 registros reconstruídos               -> 149,390 linhas


In [14]:
# --- 3.6 Winsorização dos extremos ---------------------------------------
# Substitui o corte por média + 3 desvios, que falha em distribuições assimétricas:
# os próprios outliers inflam o desvio-padrão e o limite fica acima do máximo real.
# Winsorizar (limitar) em vez de remover preserva o tamanho da amostra (ADR-006).
LIMITES = {
    'Utilizacao_Credito': ('p99',  None),   # % da linha de crédito utilizada
    'Renda_Mensal'      : ('p99',  None),
    'Razao_Divida'      : ('fixo', 2.0),    # 200% da renda — teto de negócio
}

for col, (modo, valor) in LIMITES.items():
    teto = df[col].quantile(0.99) if modo == 'p99' else valor
    acima = (df[col] > teto).sum()
    df[col] = df[col].clip(upper=teto)
    print(f'  {col:<20} teto={teto:>12,.2f}  ({modo})  {acima:>6,} valores limitados')

registrar('Extremos winsorizados', 'Utilizacao/Renda no p99, Razao com teto 2,0')

  Utilizacao_Credito   teto=        1.09  (p99)   1,494 valores limitados
  Renda_Mensal         teto=   23,084.10  (p99)   1,494 valores limitados
  Razao_Divida         teto=        2.00  (fixo)   3,418 valores limitados
Extremos winsorizados            Utilizacao/Renda no p99, Razao com teto 2,0  -> 149,390 linhas


In [15]:
# --- 3.7 Dependentes ------------------------------------------------------
moda = df['Dependentes'].mode()[0]
nulos = df['Dependentes'].isnull().sum()
df['Dependentes'] = df['Dependentes'].fillna(moda).astype(int)
registrar('Dependentes imputados', f'{nulos:,} nulos preenchidos com a moda ({moda:.0f})')

Dependentes imputados            3,828 nulos preenchidos com a moda (0)       -> 149,390 linhas


In [16]:
# --- 3.8 Verificação final da limpeza ------------------------------------
print('=' * 62)
print('VERIFICAÇÃO PÓS-LIMPEZA')
print('=' * 62)

assert df.isnull().sum().sum() == 0,        'ainda há valores nulos'
assert df['Idade'].between(18, 120).all(),  'idade fora da faixa plausível'
assert df['Razao_Divida'].max() <= 2.0,     'razão dívida/renda acima do teto'
assert set(df['Inadimplente'].unique()) <= {0, 1}, 'alvo não é binário'

retencao = len(df) / len(df_raw) * 100
print(f'Nulos restantes          : 0')
print(f'Linhas   {len(df_raw):>8,}  ->  {len(df):>8,}   (retenção {retencao:.2f}%)')
print(f'Inadimplência  {df_raw["Inadimplente"].mean()*100:.3f}%  ->  {df["Inadimplente"].mean()*100:.3f}%'
      f'   (desvio {abs(df_raw["Inadimplente"].mean()-df["Inadimplente"].mean())*100:.3f} p.p.)')
print('\nTodas as asserções passaram.')

VERIFICAÇÃO PÓS-LIMPEZA
Nulos restantes          : 0
Linhas    150,000  ->   149,390   (retenção 99.59%)
Inadimplência  6.684%  ->  6.700%   (desvio 0.016 p.p.)

Todas as asserções passaram.


---
## 4. Features analíticas

O **Score de Risco Composto** usa *rank percentil* em vez de normalização min-max.

Min-max divide pelo valor máximo observado; um único outlier extremo comprime todo o
restante da distribuição perto de zero. O rank percentil depende apenas da **posição
relativa** do cliente, o que torna o score imune a caudas longas e garante uso da
escala 0–100 inteira. Ver ADR-008.

In [17]:
# --- 4.1 Faixas categóricas ----------------------------------------------
df['Faixa_Etaria'] = pd.cut(
    df['Idade'], bins=[0, 30, 45, 60, 200],
    labels=['Jovem (18-30)', 'Adulto (31-45)', 'Maduro (46-60)', 'Sênior (60+)'])

# np.inf no limite superior: nenhuma linha cai fora da categorização.
# Faixas fechadas em 1,01 produziam nulos silenciosos para utilização > 100%.
df['Nivel_Utilizacao'] = pd.cut(
    df['Utilizacao_Credito'], bins=[-0.01, 0.25, 0.50, 0.75, np.inf],
    labels=['Baixa (0-25%)', 'Moderada (25-50%)', 'Alta (50-75%)', 'Máxima (75%+)'])

df['Nivel_Endividamento'] = pd.qcut(
    df['Razao_Divida'], q=4,
    labels=['Baixo', 'Médio', 'Alto', 'Crítico'], duplicates='drop')

for col in ['Faixa_Etaria', 'Nivel_Utilizacao', 'Nivel_Endividamento']:
    assert df[col].isnull().sum() == 0, f'{col} produziu nulos'
print('Faixas criadas sem nulos.')

Faixas criadas sem nulos.


In [18]:
# --- 4.2 Histórico de atrasos (soma ponderada pela gravidade) ------------
df['Historico_Atrasos'] = (df['Atrasos_30_59dias'] * 1 +
                           df['Atrasos_60_89dias'] * 2 +
                           df['Atrasos_90dias']    * 3)

# --- 4.3 Score de Risco Composto (0-100 por rank percentil) --------------
PESOS = {'Utilizacao_Credito': 0.40, 'Historico_Atrasos': 0.35, 'Razao_Divida': 0.25}

df['Score_Risco_Composto'] = sum(
    df[col].rank(pct=True) * peso for col, peso in PESOS.items()) * 100
df['Score_Risco_Composto'] = df['Score_Risco_Composto'].round(2)

df['Faixa_Score'] = pd.cut(
    df['Score_Risco_Composto'], bins=[0, 25, 50, 75, 100],
    labels=['Muito Baixo', 'Baixo', 'Alto', 'Muito Alto'], include_lowest=True)

print(df['Score_Risco_Composto'].describe().round(2).to_string())

count    149390.00
mean         50.00
std          17.99
min          15.69
25%          35.89
50%          49.77
75%          62.63
max          99.43


In [19]:
# --- 4.4 Validação do score ----------------------------------------------
# Um score de risco só é utilizável se a inadimplência CRESCER junto com ele.
# Sem esse teste, um score invertido passa despercebido até chegar ao painel.
decis = df.groupby(pd.qcut(df['Score_Risco_Composto'], 10, labels=False,
                           duplicates='drop'), observed=True)['Inadimplente'].mean() * 100

print('Taxa de inadimplência por decil do score:')
for d, taxa in decis.items():
    print(f'  decil {d+1:>2} | {taxa:>6.2f}%  {"#" * int(taxa)}')

monotonico = decis.is_monotonic_increasing
lift = decis.iloc[-1] / decis.iloc[0]

print(f'\nCorrelação score x inadimplência : {df["Score_Risco_Composto"].corr(df["Inadimplente"]):.4f}')
print(f'Lift (decil 10 / decil 1)        : {lift:.1f}x')
print(f'Monotônico                       : {"SIM" if monotonico else "NÃO"}')

assert monotonico, 'Score não é monotônico — não usar no painel'

Taxa de inadimplência por decil do score:
  decil  1 |   0.76%  
  decil  2 |   0.80%  
  decil  3 |   1.10%  #
  decil  4 |   1.76%  #
  decil  5 |   2.49%  ##
  decil  6 |   5.03%  #####
  decil  7 |   5.14%  #####
  decil  8 |   6.49%  ######
  decil  9 |  12.70%  ############
  decil 10 |  30.75%  ##############################



Correlação score x inadimplência : 0.3045
Lift (decil 10 / decil 1)        : 40.7x
Monotônico                       : SIM


---
## 5. KPIs

Cada agregação vira um CSV próprio. Indicadores calculados apenas para exibição no
console não chegam ao Power BI — e teriam de ser reconstruídos em DAX (ADR-010).

In [20]:
# --- 5.1 KPIs escalares ---------------------------------------------------
kpis = pd.DataFrame([
    ('Taxa Global de Inadimplência',        df['Inadimplente'].mean()*100,          '%'),
    ('Utilização Média de Crédito',         df['Utilizacao_Credito'].mean()*100,    '%'),
    ('Razão Dívida/Renda Mediana',          df['Razao_Divida'].median(),            'x'),
    ('Renda Mensal Mediana',                df['Renda_Mensal'].median(),            'USD'),
    ('Média de Linhas de Crédito',          df['Linhas_Credito_Abertas'].mean(),    'un'),
    ('Clientes com Atraso Grave (90d+)',    (df['Atrasos_90dias'] >= 1).mean()*100, '%'),
    ('Score de Risco Médio',                df['Score_Risco_Composto'].mean(),      'pts'),
    ('Lift do Score (D10/D1)',              lift,                                   'x'),
    ('Inadimplência Nacional (BCB)',        bcb['Inadimplencia_Pct'].iloc[-1],      '%'),
    ('Concessões de Crédito (BCB)',         bcb['Concessoes_Credito_RS_Mi'].iloc[-1],'R$ Mi'),
    ('Juros Médio ao Ano (BCB)',            bcb['Juros_Medio_Aa_Pct'].iloc[-1],     '% a.a.'),
    ('Correlação Juros x Inadimplência',    bcb['Juros_Medio_Aa_Pct'].corr(bcb['Inadimplencia_Pct']), 'r'),
], columns=['KPI', 'Valor', 'Unidade'])

kpis['Valor'] = kpis['Valor'].round(3)
print(kpis.to_string(index=False))

                             KPI      Valor Unidade
    Taxa Global de Inadimplência      6.700       %
     Utilização Média de Crédito     31.980       %
      Razão Dívida/Renda Mediana      0.283       x
            Renda Mensal Mediana   5246.000     USD
      Média de Linhas de Crédito      8.481      un
Clientes com Atraso Grave (90d+)      5.394       %
            Score de Risco Médio     50.000     pts
          Lift do Score (D10/D1)     40.690       x
    Inadimplência Nacional (BCB)      3.080       %
     Concessões de Crédito (BCB) 699427.000   R$ Mi
        Juros Médio ao Ano (BCB)     28.500  % a.a.
Correlação Juros x Inadimplência      0.776       r


In [21]:
# --- 5.2 KPIs agregados (tabelas de apoio do painel) ---------------------
agregados = {
    'kpi_default_por_endividamento': (
        df.groupby('Nivel_Endividamento', observed=True)
          .agg(Clientes=('Inadimplente','size'),
               Taxa_Default_Pct=('Inadimplente', lambda s: s.mean()*100),
               Renda_Media=('Renda_Mensal','mean'))),

    'kpi_default_por_faixa_etaria': (
        df.groupby('Faixa_Etaria', observed=True)
          .agg(Clientes=('Inadimplente','size'),
               Taxa_Default_Pct=('Inadimplente', lambda s: s.mean()*100),
               Renda_Media=('Renda_Mensal','mean'),
               Razao_Divida_Media=('Razao_Divida','mean'))),

    'kpi_default_por_utilizacao': (
        df.groupby('Nivel_Utilizacao', observed=True)
          .agg(Clientes=('Inadimplente','size'),
               Taxa_Default_Pct=('Inadimplente', lambda s: s.mean()*100))),

    'kpi_default_por_faixa_score': (
        df.groupby('Faixa_Score', observed=True)
          .agg(Clientes=('Inadimplente','size'),
               Taxa_Default_Pct=('Inadimplente', lambda s: s.mean()*100),
               Score_Medio=('Score_Risco_Composto','mean'))),
}

for nome, tabela in agregados.items():
    print(f'\n{nome}')
    print(tabela.round(2).to_string())


kpi_default_por_endividamento
                     Clientes  Taxa_Default_Pct  Renda_Media
Nivel_Endividamento                                         
Baixo                   37348              6.07      5919.30
Médio                   37347              5.80      7048.13
Alto                    37347              5.90      6767.23
Crítico                 37348              9.03      4908.43

kpi_default_por_faixa_etaria
                Clientes  Taxa_Default_Pct  Renda_Media  Razao_Divida_Media
Faixa_Etaria                                                               
Jovem (18-30)      10540             11.71      3148.42                0.29
Adulto (31-45)     40484              9.32      5922.47                0.41
Maduro (46-60)     53538              6.82      7049.47                0.40
Sênior (60+)       44828              3.01      6022.85                0.31

kpi_default_por_utilizacao
                   Clientes  Taxa_Default_Pct
Nivel_Utilizacao                           

In [22]:
# --- 5.3 Correlação com a inadimplência (com sinal preservado) -----------
# O sinal é a informação de negócio: idade correlaciona NEGATIVAMENTE com risco.
# Usar valor absoluto no ranking apagaria essa leitura.
VARIAVEIS = ['Utilizacao_Credito','Idade','Razao_Divida','Renda_Mensal',
             'Linhas_Credito_Abertas','Emprestimos_Imoveis','Dependentes',
             'Atrasos_30_59dias','Atrasos_60_89dias','Atrasos_90dias',
             'Historico_Atrasos','Score_Risco_Composto']

corr = (df[VARIAVEIS + ['Inadimplente']].corr(numeric_only=True)['Inadimplente']
          .drop('Inadimplente').sort_values(ascending=False)
          .rename('Correlacao').reset_index()
          .rename(columns={'index':'Variavel'}))
corr['Direcao'] = np.where(corr['Correlacao'] > 0, 'Aumenta o risco', 'Reduz o risco')
corr['Correlacao'] = corr['Correlacao'].round(4)

print(corr.to_string(index=False))

              Variavel  Correlacao         Direcao
     Historico_Atrasos      0.3934 Aumenta o risco
        Atrasos_90dias      0.3119 Aumenta o risco
  Score_Risco_Composto      0.3045 Aumenta o risco
    Utilizacao_Credito      0.2821 Aumenta o risco
     Atrasos_30_59dias      0.2715 Aumenta o risco
     Atrasos_60_89dias      0.2657 Aumenta o risco
           Dependentes      0.0465 Aumenta o risco
          Razao_Divida      0.0426 Aumenta o risco
   Emprestimos_Imoveis     -0.0076   Reduz o risco
Linhas_Credito_Abertas     -0.0305   Reduz o risco
          Renda_Mensal     -0.0522   Reduz o risco
                 Idade     -0.1157   Reduz o risco


---
## 6. Exportação para o Power BI

Todos os arquivos usam `utf-8-sig` (acentos) e ponto como separador decimal.

> **No Power Query:** defina a origem do arquivo como *Inglês (Estados Unidos)*.
> Em ambiente pt-BR, sem esse ajuste `0.766` é interpretado como `766`.

In [23]:
# Coluna técnica de auditoria não vai para o modelo do painel
COLS_EXPORT = [c for c in df.columns if c != 'Registro_Sentinela']

df[COLS_EXPORT].to_csv('output/credit_clean.csv', index=False, encoding='utf-8-sig')
kpis.to_csv('output/kpis_resumo.csv', index=False, encoding='utf-8-sig')
corr.to_csv('output/kpi_correlacoes.csv', index=False, encoding='utf-8-sig')
pd.DataFrame(auditoria).to_csv('output/log_limpeza.csv', index=False, encoding='utf-8-sig')

for nome, tabela in agregados.items():
    tabela.round(4).to_csv(f'output/{nome}.csv', encoding='utf-8-sig')

print('Arquivos gerados em output/:\n')
for arquivo in sorted(os.listdir('output')):
    caminho = os.path.join('output', arquivo)
    linhas = sum(1 for _ in open(caminho, encoding='utf-8-sig')) - 1
    print(f'  {arquivo:<36} {linhas:>7,} linhas   {os.path.getsize(caminho)/1024:>8,.1f} KB')

Arquivos gerados em output/:

  bcb_macro.csv                            120 linhas        4.9 KB
  credit_clean.csv                     149,390 linhas   16,103.4 KB
  kpi_correlacoes.csv                       12 linhas        0.5 KB
  kpi_default_por_endividamento.csv          4 linhas        0.2 KB
  kpi_default_por_faixa_etaria.csv           4 linhas        0.2 KB
  kpi_default_por_faixa_score.csv            4 linhas        0.2 KB
  kpi_default_por_utilizacao.csv             4 linhas        0.2 KB
  kpis_resumo.csv                           12 linhas        0.5 KB
  log_limpeza.csv                            7 linhas        0.5 KB


---
## 7. Resumo

| Correção | Antes | Depois |
|---|---|---|
| Séries do BCB | 2 de 3 rotuladas errado; 20714 e 25433 duplicadas | nomes oficiais conferidos; 20631 traz concessões reais |
| Score de risco | mediana 0,03; não monotônico | escala 0–100 completa; monotônico, validado por asserção |
| Códigos 96/98 | somados como contagem | tratados como sentinela |
| Razão dívida/renda | 23,4% dos valores eram dívida absoluta | reconstruída a partir da renda imputada |
| Outliers | média+3σ removia 0,8% e deixava máximos absurdos | winsorização por percentil, 99,6% de retenção |
| Nível de utilização | 2.604 nulos silenciosos | `np.inf` no limite; zero nulos |
| Duplicatas | não verificadas | 609 removidas |
| Exportação | 9 KPIs escalares | 9 arquivos, incluindo agregados e log de limpeza |

As decisões e seus motivos estão em [`docs/ADR.md`](../docs/ADR.md).